pkg/dockerfile/expand.go

```diff
@@ -16,20 +16,62 @@ func (d *Dockerfile) Expand(env SingleWordExpander) {
 
 func (d *Dockerfile) expand(env instructions.SingleWordExpander) {
 	metaArgsEnvExpander := d.metaArgsEnvExpander(env)
+
+	// Create an os.Expand-compatible function from the SingleWordExpander
+	expandFunc := func(key string) string {
+		value, err := metaArgsEnvExpander(key)
+		if err != nil {
+			return ""
+		}
+		return value
+	}
+
 	for i, stage := range d.Stages {
-		d.Stages[i].BaseName = os.Expand(stage.BaseName, func(key string) string {
-			value, err := metaArgsEnvExpander(key)
-			if err != nil {
-				return ""
-			}
-			return value
-		})
+		d.Stages[i].BaseName = os.Expand(stage.BaseName, expandFunc)
+
 		for i := range stage.Commands {
-			cmdExpander, ok := stage.Commands[i].Command.(instructions.SupportsSingleWordExpansion)
-			if ok {
-				cmdExpander.Expand(metaArgsEnvExpander)
```

Comment:
So, I did some digging. The previous approach isn't necessarily wrong. All the instructions define how they should be expanded, so calling their `Expand()` method is probably better than re-implementing it like we do in this PR.

The problem is that the `metaArgsEnvExpander` function is completely wrong 😞